kernel/src/events/subject.rs

```diff
@@ -70,19 +70,28 @@ impl<E: Events, F: EventsFilter<E>> Subject<E, F> {
         }
 
         // Slow path: broadcast the new events to all observers.
+        let mut active_observers = Vec::new();
+        let mut num_freed = 0;
         let mut observers = self.observers.lock();
         observers.retain(|observer, filter| {
             if let Some(observer) = observer.upgrade() {
-                if !filter.filter(events) {
-                    return true;
+                if filter.filter(events) {
+                    active_observers.push(observer.clone());
                 }
-                observer.on_events(events);
                 true
             } else {
-                self.num_observers.fetch_sub(1, Ordering::Relaxed);
+                num_freed += 1;
                 false
             }
         });
+        if num_freed > 0 {
+            self.num_observers.fetch_sub(num_freed, Ordering::Relaxed);
+        }
+        drop(observers);
+
+        for observer in active_observers {
+            observer.on_events(events);
```

Comment:
Yes, you are right, in theory this optimization is negative. I weakly argue that by some simple tests and code review, most of the time we have one or zero observer on the file, which will make this optimization benefited. More than one observers appears in a rather smaller scope: `epoll`, sockests, etc.

Reply:
Could you analyze the number of observers in case of epoll or sockets? If all use cases in practice have a few observers, then this approach is acceptable, at least for now. If not, we might have to think of an alternative approach.

Reply:
I look through the code that it seems only `epoll_ctl` would add multiple observers on a same `FileTableEntry`, most of the time there is only one or none.

And I did a small test to count and evaluate the impact of this PR for some suspicious cases, showed as below:

| micro benchmark               | max number of triggered observers | without this PR (micros) | have this PR (micros) |
| ----------------------------- | --------------------------------- | ------------------------ | --------------------- |
| lmbench-open                  | 1                                 | 1.1874                   | 1.1408                |
| lmbench-write                 | 1                                 | 0.2046                   | 0.1757                |
| lmbench-select-file           | 1                                 | 2.0160                   | 2.0011                |
| lmbench-pipe-latency          | 1                                 | 3.1888                   | 3.0501                |
| lmbench-select-tcp            | 1                                 | 1.7950                   | 1.7410                |
| lmbench-unix-latency          | 1                                 | 3.8558                   | 3.7089                |
| lmbench-tcp-latency-localhost | 1                                 | 20.0865                  | 19.1732               |
|                               |                                   |                          |                       |
| gvisor_syscall_test           | 2                                 |                          |                       |

It seems this PR brought benefits most of the time for now.



And I also think of an alternative approach which needs more refactoring work. Currently we add a `SpinLock` to the whole `FileTable`, which contains a `table: SlotVec<FileTableEntry>` and a `subject: Subject<FdEvents>`, actually we can make the locking granularity finer, that decouples the `table` and the `subject`, and the `Subject` within `FileTableEntry` can also be operated while we donot hold spin-locks, since the `FdEvents` only cares when we remove some entries, this way we can make the operation on `table` and `subject` independently. And this will require modifying every usage locations of using the `FileTable` now.

Reply:
Good. Then I think the current approach is acceptable, at least for now.